File: src/scanners/dive.ts
import type { IScanner, ImageSource, ScannerResult } from '../types/index.js';
import { execWithTimeout, getToolVersion, isToolAvailable, parseJsonFile, writeFallbackResult, createTimedResult } from './utils.js';

const TIMEOUT_MS = 240_000;

function buildCommand(source: ImageSource, outputPath: string): string {
  switch (source.type) {
    case 'docker':
      return `dive "${source.ref}" --json "${outputPath}"`;
    case 'tar':
      return `dive --source docker-archive "${source.path}" --json "${outputPath}"`;
    case 'registry':
      throw new Error('Dive does not support direct registry scanning');
  }
}

export class DiveScanner implements IScanner {
  readonly name = 'dive';

  async scan(source: ImageSource, outputPath: string): Promise<ScannerResult> {
    const start = Date.now();
    try {
      const command = buildCommand(source, outputPath);
      await execWithTimeout(command, TIMEOUT_MS);

      const data = await parseJsonFile(outputPath);
      return createTimedResult(this.name, true, start, { data });
    } catch (error) {
      const message = error instanceof Error ? error.message : String(error);
      console.warn('Dive scan failed:', message);
      await writeFallbackResult(outputPath, message, { layer: [] });
      return createTimedResult(this.name, false, start, { error: message });
    }
  }

  async getVersion(): Promise<string> {
    return getToolVersion('dive --version');
  }

  async isAvailable(): Promise<boolean> {
    return isToolAvailable('dive');
  }

  supportsSource(source: ImageSource): boolean {
    return source.type !== 'registry';
  }
}

File: internal/scanner/dive.go
package scanner

import (
	"context"
	"fmt"
	"os"
	"time"

	"github.com/HarborGuard/harborguard-sensor/internal/types"
)

const diveTimeoutMs = 240000

type DiveScanner struct{}

func (d *DiveScanner) Name() string { return "dive" }

func (d *DiveScanner) Scan(ctx context.Context, source types.ImageSource, outputPath string) (*types.ScannerResult, error) {
	start := time.Now()

	if source.Type == "registry" {
		msg := "Dive does not support direct registry scanning"
		durationMs := time.Since(start).Milliseconds()
		return &types.ScannerResult{Scanner: "dive", Success: false, Error: msg, DurationMs: durationMs}, nil
	}

	cmd := d.buildCommand(source, outputPath)
	_, _, err := ExecWithTimeout(ctx, cmd, diveTimeoutMs, nil)
	durationMs := time.Since(start).Milliseconds()

	if err != nil {
		msg := err.Error()
		fmt.Fprintf(os.Stderr, "Dive scan failed: %s\n", msg)
		_ = WriteFallbackResult(outputPath, msg, map[string]interface{}{"layer": []interface{}{}})
		return &types.ScannerResult{Scanner: "dive", Success: false, Error: msg, DurationMs: durationMs}, nil
	}

	var data interface{}
	if parseErr := ParseJSONFile(outputPath, &data); parseErr != nil {
		msg := parseErr.Error()
		return &types.ScannerResult{Scanner: "dive", Success: false, Error: msg, DurationMs: durationMs}, nil
	}

	return &types.ScannerResult{Scanner: "dive", Success: true, Data: data, DurationMs: durationMs}, nil
}

func (d *DiveScanner) buildCommand(source types.ImageSource, outputPath string) string {
	switch source.Type {
	case "tar":
		return fmt.Sprintf(`dive --source docker-archive "%s" --json "%s"`, source.Path, outputPath)
	default:
		return fmt.Sprintf(`dive "%s" --json "%s"`, source.Ref, outputPath)
	}
}

func (d *DiveScanner) GetVersion() string {
	return GetToolVersion("dive --version")
}

func (d *DiveScanner) IsAvailable() bool {
	return IsToolAvailable("dive")
}

func (d *DiveScanner) SupportsSource(source types.ImageSource) bool {
	return source.Type != "registry"
}
